src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2022 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::fs::DirBuilder;
+use std::io::{Error, ErrorKind, Result};
+use std::os::unix::fs::DirBuilderExt;
+use std::path::{Path, PathBuf};
+
+use crate::{safe_join, SafePathBuf};
+
+const DIRECTORY_MODE_DEFAULT: u32 = 0o700;
+const DIRECTORY_MODE_MASK: u32 = 0o777;
+
+/// Safe version of `DirBuilder` to protect from TOCTOU style of attacks.
+#[derive(Debug)]
+pub struct SafeDirBuilder {
+    root: PathBuf,
+    mode: u32,
+    recursive: bool,
+}
+
+impl SafeDirBuilder {
+    /// Creates a new set of options with default mode/security settings for all platforms and
+    /// also non-recursive.
+    pub fn new<P: AsRef<Path>>(root: P) -> Result<Self> {
+        let root = root.as_ref().canonicalize()?;
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        Ok(SafeDirBuilder {
+            root,
+            mode: DIRECTORY_MODE_DEFAULT,
+            recursive: false,
+        })
+    }
+
+    /// Indicates that directories should be created recursively, creating all parent directories.
+    ///
+    /// Parents that do not exist are created with the same security and permissions settings.
+    pub fn recursive(&mut self) -> &mut Self {
+        self.recursive = true;
+        self
+    }
+
+    /// Sets the mode to create new directories with. This option defaults to 0o755.
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode & DIRECTORY_MODE_MASK;
+        self
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `SafePathBuf::root()`, otherwise error will be returned.
+    /// It is considered an error if the directory already exists unless recursive mode is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<SafePathBuf> {
+        let mut root = self.root.clone();
+        let path = safe_join("/", path)?;
```

Comment:
The new implementation should help to understand the usage:
```
    /// Creates the specified directory with the options configured in this builder.
    ///
    /// The `path` must be a subdirectory of `SafeDirBuilder::root`, otherwise error will be
    /// returned. It is considered an error if the directory already exists unless recursive mode
    /// is enabled.
    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<PinnedPathBuf> {
        let path = scoped_join("/", path)?;
        self.do_mkdir(&path)
    }

    /// Creates sub-directory with the options configured in this builder.
    ///
    /// It is considered an error if the directory already exists unless recursive mode is enabled.
    pub fn create_sub_directory<P: AsRef<Path>>(&self, path: P) -> Result<PinnedPathBuf> {
        let path = scoped_join(&self.root, path)?;
        self.do_mkdir(&path)
    }
```